// initi-loader/index.js
import setupDom from "./setup/setupDOM.js";
import initRequirejs from "./setup/requirejs.js";
import initLoader from "./loader/main.js";
import createGlobalLoaderMethods from "./setup/loaderMethods.js";

const config = {};

export async function init(options) {
    Object.assign(config, options);
    setupDom();
    const requirejs = initRequirejs({ paths: { guiCode: options.paths.guiCode } });
    const { loader, Loader } = initLoader({ 
        loaderConfig: options.loaderOptions,
        loaderPath: options.paths.loader,
    });
    global.loader = globalThis.loader = loader;
    global.Loader = globalThis.Loader = Loader;

    requirejs.config({
        baseUrl: options.paths.guiCode,
        paths: {
            css: "require/css",
            json: "require/json",
            text: "require/text",
            styles: "../styles",
            jsonFiles: "../json",
            d3: "external/d3",
            loader: "external/loader/requirejsmodule",
        },
        loaderInstance: global.loader,
    });

    await createGlobalLoaderMethods();
}

export const version = "1.0.0";
